### PR TITLE
Update reading DICOM and marker visualization

### PR DESCRIPTION
**Pull request by sz144**
### Description
- New feature to separate different dicom series by SeriesInstanceUID
- Enhancement: marker visualization changed to `plt.scatter()` from `plt.plot()`
- Data used in CMR example changed to ShefPAH-179 v2.0 
- Add README to examples of `autism_detection` and `toy_domain_adapation`
- Two badges added in README.md, with two removed and all reordered
- New tests to cover the changes
- MNIST-M server switched back to github as zenodo is too slow and often causes test errors
- Removed zenodo jason (not working actually)

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: examples/cmri_mpca/main.py [hunk 1]
```diff
@@ -70,43 +70,69 @@ def main():
...
+
     if save_images:
-        visualize.plot_multi_images(
-            [images[i][0, ...] for i in range(n_samples)],
-            marker_locs=landmarks,
-            im_kwargs=dict(cfg.IM_KWARGS),
-            marker_kwargs=dict(cfg.MARKER_KWARGS),
-            n_cols=10,
-        ).savefig(str(save_images_location) + "/0)first_phase.png")
+        n_img_per_fig = 35
```
Comment (haipinglu) [inline examples/cmri_mpca/main.py:83]: Is 35 always the best number: would you consider make this part of the configuration in config yaml?

File: examples/cmri_mpca/main.py [hunk 2]
```diff
@@ -70,43 +70,69 @@ def main():
...
+        for k in range(n_figures):
+            visualize.plot_multi_images(
+                [images[i][0, ...] for i in range(k * n_img_per_fig, min((k + 1) * n_img_per_fig, n_samples))],
+                marker_locs=landmarks[k * n_img_per_fig : min((k + 1) * n_img_per_fig, n_samples), :],
+                im_kwargs=dict(cfg.IM_KWARGS),
+                marker_cmap="Set1",
+                marker_kwargs=dict(cfg.MARKER_KWARGS),
+                marker_titles=markers,
+                image_titles=list(patient_ids[k * n_img_per_fig : min((k + 1) * n_img_per_fig, n_samples)]),
+                n_cols=5,
```
Comment (haipinglu) [inline examples/cmri_mpca/main.py:94]: Again, have you consider 5 here as config rather than hard code?

File: examples/cmri_mpca/main.py [hunk 3]
```diff
@@ -70,43 +70,69 @@ def main():
...
 
     # ---- data pre-processing ----
     # ----- image registration -----
     img_reg, max_dist = reg_img_stack(images.copy(), landmarks, landmarks[0])
     if save_images:
         visualize.plot_multi_images(
-            [img_reg[i][0, ...] for i in range(n_samples)], im_kwargs=dict(cfg.IM_KWARGS), n_cols=10
+            [img_reg[i][0, ...] for i in range(n_samples)],
+            im_kwargs=dict(cfg.IM_KWARGS),
+            n_cols=10,
```
Comment (haipinglu) [inline examples/cmri_mpca/main.py:104]: Again, consider hard code vs yaml config

File: examples/cmri_mpca/main.py [hunk 4]
```diff
@@ -70,43 +70,69 @@ def main():
...
         ).savefig(str(save_images_location) + "/1)image_registration")
 
     # ----- masking -----
     img_masked = mask_img_stack(img_reg.copy(), mask)
     if save_images:
         visualize.plot_multi_images(
-            [img_masked[i][0, ...] for i in range(n_samples)], im_kwargs=dict(cfg.IM_KWARGS), n_cols=10
+            [img_masked[i][0, ...] for i in range(n_samples)],
+            im_kwargs=dict(cfg.IM_KWARGS),
+            n_cols=10,
```
Comment (haipinglu) [inline examples/cmri_mpca/main.py:114]: If you are using the same para multiple times, config seems better in the long term.

File: examples/cmri_mpca/main.py [hunk 5]
```diff
@@ -70,43 +70,69 @@ def main():
...
         ).savefig(str(save_images_location) + "/2)masking")
 
     # ----- resize -----
     img_rescaled = rescale_img_stack(img_masked.copy(), scale=1 / cfg.PROC.SCALE)
     if save_images:
         visualize.plot_multi_images(
-            [img_rescaled[i][0, ...] for i in range(n_samples)], im_kwargs=dict(cfg.IM_KWARGS), n_cols=10
+            [img_rescaled[i][0, ...] for i in range(n_samples)],
+            im_kwargs=dict(cfg.IM_KWARGS),
+            n_cols=10,
```
Comment (haipinglu) [inline examples/cmri_mpca/main.py:124]: If you are using the same para multiple times, config seems better in the long term.

File: examples/cmri_mpca/main.py [hunk 6]
```diff
@@ -70,43 +70,69 @@ def main():
...
         ).savefig(str(save_images_location) + "/3)resize")
 
     # ----- normalization -----
     img_norm = normalize_img_stack(img_rescaled.copy())
     if save_images:
         visualize.plot_multi_images(
-            [img_norm[i][0, ...] for i in range(n_samples)], im_kwargs=dict(cfg.IM_KWARGS), n_cols=10
+            [img_norm[i][0, ...] for i in range(n_samples)],
+            im_kwargs=dict(cfg.IM_KWARGS),
+            n_cols=10,
```
Comment (haipinglu) [inline examples/cmri_mpca/main.py:134]: If you are using the same para multiple times, config seems better in the long term.

File: examples/cmri_mpca/tutorial.ipynb [hunk 7]
```diff
@@ -197,16 +199,21 @@
...
-    "    marker_locs=landmarks,\n",
-    "    im_kwargs=dict(cfg.IM_KWARGS),\n",
-    "    marker_titles=markers,\n",
-    "    marker_cmap=\"Set1\",\n",
-    "    marker_kwargs=dict(cfg.MARKER_KWARGS),\n",
-    "    n_cols=10\n",
-    ").show()"
+    "# plot the first phase of images with landmarks\n",
+    "\n",
+    "n_img_per_fig = 35\n",
```
Comment (haipinglu) [inline examples/cmri_mpca/tutorial.ipynb:204]: If 35 is the default config, specifying so in config.py may be better.

File: examples/cmri_mpca/tutorial.ipynb [hunk 8]
```diff
@@ -197,16 +199,21 @@
...
+    "for k in range(n_figures):\n",
+    "    visualize.plot_multi_images(\n",
+    "        [images[i][0, ...] for i in range(k * n_img_per_fig, min((k + 1) * n_img_per_fig, n_samples))],\n",
+    "        marker_locs=landmarks[k * n_img_per_fig: min((k + 1) * n_img_per_fig, n_samples), :],\n",
+    "        im_kwargs=dict(cfg.IM_KWARGS),\n",
+    "        marker_cmap=\"Set1\",\n",
+    "        marker_kwargs=dict(cfg.MARKER_KWARGS),\n",
+    "        marker_titles=markers,\n",
+    "        image_titles=list(patient_ids[k * n_img_per_fig: min((k + 1) * n_img_per_fig, n_samples)]),\n",
+    "        n_cols=5,\n",
```
Comment (haipinglu) [inline examples/cmri_mpca/tutorial.ipynb:215]: consider to config or hardcode

File: examples/toy_domain_adaptation/README.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# Domain Adaptation on Toy Data
+
+### 1. Description
+
+This example demonstrates domain adaptation methods random generated toy data.
```
Comment (haipinglu) [inline examples/toy_domain_adaptation/README.md:5]: methods random generated --> methods on randomly generated

File: kale/interpret/visualize.py [hunk 10]
```diff
@@ -78,6 +79,7 @@ def plot_multi_images(
             Defaults to None.
         im_kwargs (dict, optional): Key word arguments for plotting images. Defaults to None.
```
Comment (haipinglu) [inline kale/interpret/visualize.py:80]: These three lines: keyword or key word? Google to confirm.

File: kale/loaddata/image_access.py [hunk 11]
```diff
@@ -359,7 +359,36 @@ def read_dicom_phases(dicom_path, sort_instance=True):
...
+            dcms_out.append(dcms[uid])
+    else:
+        dcms_out = [dcm_phases]
+
+    return dcms_out
+
+
+def read_dicom_dir(dicom_path, sort_instance=True, sort_patient=False, check_series_uid=False):
     """Read dicom files for multiple patients and multiple instances / phases from a given directory arranged in the
         following structure: ::
```
Comment (haipinglu) [inline kale/loaddata/image_access.py:393]: why three colons : :: ?

**Review by haipinglu — CHANGES_REQUESTED** [review]
Good job. Some minor comments to address and then we can merge.

File: examples/cmri_mpca/tutorial.ipynb (hunk 7 above)
Comment (sz144) [reply examples/cmri_mpca/tutorial.ipynb:204]: No, specifically in this example

File: examples/toy_domain_adaptation/README.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
+# Domain Adaptation on Toy Data
+
+### 1. Description
+
+This example demonstrates domain adaptation methods on random generated toy data.
```
Comment (haipinglu) [inline examples/toy_domain_adaptation/README.md:5]: random --> randomly !

Comment (haipinglu) [reply examples/toy_domain_adaptation/README.md:5]: random is describing generate (verb) so needs to be an adv NOT adj

File: examples/cmri_mpca/tutorial.ipynb (hunk 8 above)
Comment (haipinglu) [reply examples/cmri_mpca/tutorial.ipynb:215]: @sz144 so what's the response? (two places)

File: examples/cmri_mpca/tutorial.ipynb (hunk 7 above)
Comment (haipinglu) [reply examples/cmri_mpca/tutorial.ipynb:204]: Why not configure it? so users can just modify the config file?

File: examples/cmri_mpca/main.py [hunk 13]
```diff
@@ -70,44 +71,64 @@ def main():
...
+    marker_names = list(landmark_df.columns[1::2])
+    markers = []
+    for marker in marker_names:
+        marker_name = marker.split(" ")
+        marker_name.pop(-1)
+        marker_name = " ".join(marker_name)
+        markers.append(marker_name)
+
+    if save_figs:
+        n_img_per_fig = 45
```
Comment (haipinglu) [inline examples/cmri_mpca/main.py:84]: Why not configure it? so users can just modify the config file?

File: examples/cmri_mpca/main.py [hunk 14]
```diff
@@ -70,44 +71,64 @@ def main():
...
+        for k in range(n_figures):
+            visualize.plot_multi_images(
+                [images[i][0, ...] for i in range(k * n_img_per_fig, min((k + 1) * n_img_per_fig, n_samples))],
+                marker_locs=landmarks[k * n_img_per_fig : min((k + 1) * n_img_per_fig, n_samples), :],
+                im_kwargs=dict(cfg.IM_KWARGS),
+                marker_cmap="Set1",
+                marker_kwargs=dict(cfg.MARKER_KWARGS),
+                marker_titles=markers,
+                image_titles=list(patient_ids[k * n_img_per_fig : min((k + 1) * n_img_per_fig, n_samples)]),
+                n_cols=5,
```
Comment (haipinglu) [inline examples/cmri_mpca/main.py:95]: Again, have you consider 5 here as config rather than hard code?

**Review by haipinglu — COMMENTED** [review]
Need to talk in person to resolve the remaining issues to avoid a loop

File: examples/cmri_mpca/tutorial.ipynb [hunk 15]
```diff
@@ -2,15 +2,23 @@
  "cells": [
   {
    "cell_type": "markdown",
-   "metadata": {},
+   "metadata": {
+    "pycharm": {
```
Comment (haipinglu) [inline examples/cmri_mpca/tutorial.ipynb:6]: Find a way to remove this consistently to reduce review burden and show changes more clearly.

**Review by haipinglu — CHANGES_REQUESTED** [review]
We can merge after removing the unnecessary pycharm metadata key info in notebook(s).

**Review by haipinglu — APPROVED** [review]

**Pull request merged**